### PR TITLE
use new bigdata-operator as notebook session backend

### PR DESCRIPTION
**Pull request by ImpSy**
## Why

Remove the need for NFS in the notebook-service

## What

- Add a new API in bigdata-operator for remoteKernelSession
- Make the notebook-service use webhookKernelSessionManager w/ the bigdata-operator

### API server started
```
2024-04-19T14:33:06Z    INFO    setup    bigdata-operator    {"buildVersion": "0.4.11-429c6a8f", "buildDate": "2024-04-19T13:34:26Z"}
2024-04-19T14:33:06Z    INFO    controller-runtime.metrics    Metrics server is starting to listen    {"addr": ":8080"}
2024-04-19T14:33:06Z    INFO    spotConfigProvider    Found Ocean controller namespace "kube-system"
2024-04-19T14:33:06Z    INFO    setup    oceanControllerNamespace: "kube-system"
2024-04-19T14:33:06Z    INFO    spotConfigProvider    SPOTINST_CLUSTER_IDENTIFIER found in config map kube-system/spotinst-kubernetes-cluster-controller-config
2024-04-19T14:33:06Z    INFO    setup    controllerClusterID: "privatelink-india"
2024-04-19T14:33:06Z    INFO    spotConfigProvider    SPOTINST_BASE_URL found in env var
2024-04-19T14:33:06Z    INFO    spotConfigProvider    SPOTINST_TOKEN found in secret kube-system/spotinst-kubernetes-cluster-controller
2024-04-19T14:33:06Z    INFO    spotConfigProvider    SPOTINST_ACCOUNT found in secret kube-system/spotinst-kubernetes-cluster-controller
2024-04-19T14:33:06Z    INFO    setup    starting API server
2024-04-19T14:33:06Z    INFO    setup    starting manager
2024-04-19T14:33:06Z    INFO    Starting server    {"path": "/metrics", "kind": "metrics", "addr": ":8080"}
2024-04-19T14:33:06Z    INFO    Starting server    {"kind": "health probe", "addr": ":8082"}
I0419 14:33:06.997726       1 leaderelection.go:248] attempting to acquire leader lease spot-system/6aaa1a16.spot.io...
I0419 14:33:22.775233       1 leaderelection.go:258] successfully acquired lease spot-system/6aaa1a16.spot.io
```

### Notebook-service using bigdata-operator as backend
```
Starting Jupyter Enterprise Gateway...
[D 2024-04-19 14:45:58.877 EnterpriseGatewayApp] Searching ['/venv/etc/jupyter', '/app/services/bigdata-notebook-service/.jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 2024-04-19 14:45:58.878 EnterpriseGatewayApp] Looking for jupyter_config in /etc/jupyter
[D 2024-04-19 14:45:58.878 EnterpriseGatewayApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 2024-04-19 14:45:58.878 EnterpriseGatewayApp] Looking for jupyter_config in /app/services/bigdata-notebook-service/.jupyter
[D 2024-04-19 14:45:58.879 EnterpriseGatewayApp] Looking for jupyter_config in /venv/etc/jupyter
[D 2024-04-19 14:45:58.880 EnterpriseGatewayApp] Looking for jupyter_enterprise_gateway_config in /etc/jupyter
[D 2024-04-19 14:45:58.880 EnterpriseGatewayApp] Looking for jupyter_enterprise_gateway_config in /usr/local/etc/jupyter
[D 2024-04-19 14:45:58.880 EnterpriseGatewayApp] Looking for jupyter_enterprise_gateway_config in /app/services/bigdata-notebook-service/.jupyter
[D 2024-04-19 14:45:58.881 EnterpriseGatewayApp] Looking for jupyter_enterprise_gateway_config in /venv/etc/jupyter
[I 2024-04-19 14:45:58.883 EnterpriseGatewayApp] Initializing BigdataKernelSpecManager
[I 2024-04-19 14:45:58.885 EnterpriseGatewayApp] Webhook kernel session persistence activated
[I 2024-04-19 14:45:58.894 EnterpriseGatewayApp] Jupyter Enterprise Gateway 3.2.2 is available at http://0.0.0.0:8888
[I 240419 14:46:23 web:2348] 200 GET /api (127.0.0.1) 0.85ms
```

```
{"level":"info","service":"kernel-session-manager","httpRequest":{"proto":"HTTP/1.1","remoteIP":"192.168.197.60:56762","requestID":"bigdata-operator-6f684d75bf-56g8c/BtBoDD0yso-000007","requestMethod":"GET","requestPath":"/kernelSession","requestURL":"http://bigdata-operator-api.spot-system.svc.cluster.local/kernelSession"},"httpRequest":{"header":{"accept":"*/*","accept-encoding":"gzip, deflate","connection":"keep-alive","user-agent":"python-requests/2.31.0"},"proto":"HTTP/1.1","remoteIP":"192.168.197.60:56762","requestID":"bigdata-operator-6f684d75bf-56g8c/BtBoDD0yso-000007","requestMethod":"GET","requestPath":"/kernelSession","requestURL":"http://bigdata-operator-api.spot-system.svc.cluster.local/kernelSession","scheme":"http"},"timestamp":"2024-04-19T14:46:53.8236137Z","message":"Request: GET /kernelSession"}
{"level":"info","service":"kernel-session-manager","sessions_found":0,"timestamp":"2024-04-19T14:46:53.827606353Z","message":"list kernel sessions"}
{"level":"info","service":"kernel-session-manager","httpRequest":{"proto":"HTTP/1.1","remoteIP":"192.168.197.60:56762","requestID":"bigdata-operator-6f684d75bf-56g8c/BtBoDD0yso-000007","requestMethod":"GET","requestPath":"/kernelSession","requestURL":"http://bigdata-operator-api.spot-system.svc.cluster.local/kernelSession"},"httpResponse":{"bytes":2,"elapsed":3.990853,"header":{"content-type":"application/json"},"status":200},"timestamp":"2024-04-19T14:46:53.827639366Z","message":"Response: 200 OK"}
```

### CODE REVIEW

File: charts/bigdata-notebook-service/Chart.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ apiVersion: v2
 name: bigdata-notebook-service
 description: A Helm chart for the Spot Big Data Notebook Service
 type: application
-version: 0.2.7
+version: 0.3.0
 appVersion: 0.1.16
```
Comment (thorsteinnth) [inline charts/bigdata-notebook-service/Chart.yaml:6]: need to bump bns image and appVersion?

File: charts/bigdata-operator/Chart.yaml [hunk 2]
```diff
@@ -2,7 +2,7 @@ apiVersion: v2
 name: bigdata-operator
 description: Spot Ocean BigData Operator
 type: application
-version: 0.4.13
+version: 0.4.14
 appVersion: 0.4.10
```
Comment (thorsteinnth) [inline charts/bigdata-operator/Chart.yaml:6]: need to bump image and app version

File: charts/bigdata-notebook-service/templates/deployment.yaml [hunk 3]
```diff
@@ -113,12 +91,16 @@ spec:
             value: "{{ sub .Values.kernel.portMax .Values.kernel.portMin }}"
           - name: EG_KERNEL_SESSION_PERSISTENCE
             value: "True"
+          - name: EG_AVAILABILITY_MODE
```
Comment (thorsteinnth) [inline charts/bigdata-notebook-service/templates/deployment.yaml:94]: what's this?

Comment (ImpSy) [reply charts/bigdata-notebook-service/templates/deployment.yaml:94]: Without this env var we have this line in the log
```
[I 2024-04-12 13:51:26.527 EnterpriseGatewayApp] Kernel session persistence is enabled but availability mode is not.  Setting EnterpriseGatewayApp.availability_mode to 'replication'.
```
added it to remove the noise

File: charts/bigdata-notebook-service/templates/deployment.yaml [hunk 4]
```diff
@@ -94,7 +72,7 @@ spec:
           - name: EG_NAMESPACE
             value: {{ .Release.Namespace }}
           - name: EG_LIST_KERNELS
-            value: {{ .Values.eglistKernels }}
+            value: {{ .Values.egListKernels | quote }}
```
Comment (thorsteinnth) [inline charts/bigdata-notebook-service/templates/deployment.yaml:75]: do we need this quote? already a quoted string in values.yaml?

File: charts/bigdata-operator/templates/api-service.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ include "bigdata-operator.fullname" . }}-api
```
Comment (thorsteinnth) [inline charts/bigdata-operator/templates/api-service.yaml:5]:
```suggestion
  name: bigdata-operator-api
```

Comment (thorsteinnth) [reply charts/bigdata-operator/templates/api-service.yaml:5]: since this is hardcoded in the bns chart, hardcode it here as well?

Comment (ImpSy) [reply charts/bigdata-operator/templates/api-service.yaml:5]: didn't wanted to hardcode it in case we ever do blue green

**Review by thorsteinnth — APPROVED** [review]

**Pull request merged**